third_party/terraform/utils/common_polling.go

```diff
@@ -31,19 +32,75 @@ func SuccessPollResult() PollResult {
 	return nil
 }
 
-func PollingWaitTime(pollF PollReadFunc, checkResponse PollCheckResponseFunc, activity string, timeout time.Duration) error {
+func PollingWaitTime(pollF PollReadFunc, checkResponse PollCheckResponseFunc, activity string,
+	timeout time.Duration, targetOccurrences int) error {
 	log.Printf("[DEBUG] %s: Polling until expected state is read", activity)
-	return resource.Retry(timeout, func() *resource.RetryError {
+	log.Printf("[DEBUG] Target occurrences: %d", targetOccurrences)
+	return RetryWithTargetOccurrences(timeout, targetOccurrences, func() *resource.RetryError {
 		readResp, readErr := pollF()
 		return checkResponse(readResp, readErr)
 	})
 }
 
+// RetryWithTargetOccurrences is a basic wrapper around StateChangeConf that will retry
+// a function until it returns the specified amount of target occurrences continuously.
+// Adapted from the Retry function in the go SDK.
+func RetryWithTargetOccurrences(timeout time.Duration, targetOccurrences int,
+	f resource.RetryFunc) error {
+	// These are used to pull the error out of the function; need a mutex to
+	// avoid a data race.
+	var resultErr error
+	var resultErrMu sync.Mutex
+
+	c := &resource.StateChangeConf{
+		Pending:                   []string{"retryableerror"},
+		Target:                    []string{"success"},
+		Timeout:                   timeout,
+		MinTimeout:                500 * time.Millisecond,
+		ContinuousTargetOccurence: targetOccurrences,
+		Refresh: func() (interface{}, string, error) {
+			rerr := f()
+
+			resultErrMu.Lock()
+			defer resultErrMu.Unlock()
+
+			if rerr == nil {
+				resultErr = nil
+				return 42, "success", nil
+			}
+
+			resultErr = rerr.Err
+
+			if rerr.Retryable {
+				return 42, "retryableerror", nil
+			}
+			return nil, "quit", rerr.Err
+		},
+	}
+
+	_, waitErr := c.WaitForState()
+
+	// Need to acquire the lock here to be able to avoid race using resultErr as
```

Comment:
How is this true? When we WaitForState why would there be any chance that resultErr hasn't been set to the correct value prior to returning?

I'm not super familiar with how WaitForState works, so it could be possible